cmd/upgrade.go

```diff
@@ -176,53 +183,72 @@ func (uc *upgradeCmd) loadCluster(cmd *cobra.Command) error {
 		return errors.Errorf("upgrading from Kubernetes version %s to version %s is not supported. To see a list of available upgrades, use 'aks-engine get-versions --version %s'", uc.containerService.Properties.OrchestratorProfile.OrchestratorVersion, uc.upgradeVersion, uc.containerService.Properties.OrchestratorProfile.OrchestratorVersion)
 	}
+	return nil
+}
 
-	// Read the name suffix from the parameters to identify VMs in the resource group that belong to this cluster.
-	templatePath := path.Join(uc.deploymentDirectory, "azuredeploy.json")
-	contents, err := ioutil.ReadFile(templatePath)
-	if err != nil {
-		return errors.Wrap(err, "error reading ARM file")
+func (uc *upgradeCmd) validateCurrentLocalState(armTemplateHandle io.Reader) error {
+	if uc.containerService.Location == "" {
+		uc.containerService.Location = uc.location
+	} else if uc.containerService.Location != uc.location {
+		return errors.New("--location does not match api model location")
 	}
 
-	var template interface{}
-	json.Unmarshal(contents, &template)
+	if !uc.force {
+		err := uc.validateTargetVersion()
+		if err != nil {
+			return errors.Wrap(err, "Invalid upgrade target version. Consider using --force if you really want to proceed")
+		}
+	}
 
-	var templateMap, templateParameters, nameSuffixParam map[string]interface{}
-	var okType bool
+	//allows to identify VMs in the resource group that belong to this cluster.
+	if nameSuffix, err := readNameSuffixFromArmTemplate(armTemplateHandle); err == nil {
+		uc.nameSuffix = nameSuffix
+	} else {
+		return errors.Wrap(err, "Failed to read nameSuffix from azuredeploy.json")
+	}
+
+	log.Infoln(fmt.Sprintf("Upgrading cluster with name suffix: %s", uc.nameSuffix))
+
+	uc.agentPoolsToUpgrade = make(map[string]bool)
+	uc.agentPoolsToUpgrade[kubernetesupgrade.MasterPoolName] = true
+	for _, agentPool := range uc.containerService.Properties.AgentPoolProfiles {
+		uc.agentPoolsToUpgrade[agentPool.Name] = true
+	}
+	return nil
+}
 
-	if templateMap, okType = template.(map[string]interface{}); !okType {
-		return errors.Errorf("error asserting data from file %q", templatePath)
+func readNameSuffixFromArmTemplate(armTemplateHandle io.Reader) (string, error) {
+	var template *map[string]interface{}
+	decoder := json.NewDecoder(armTemplateHandle)
+	if err := decoder.Decode(&template); err != nil {
+		return "", err
 	}
 
+	var templateParameters, nameSuffixParam map[string]interface{}
+	var okType bool
+
 	const (
 		parametersKey   = "parameters"
 		nameSuffixKey   = "nameSuffix"
 		defaultValueKey = "defaultValue"
 	)
 
-	if templateParameters, okType = templateMap[parametersKey].(map[string]interface{}); !okType {
-		return errors.Errorf("error asserting data from key \"%s\" in file %q",
-			parametersKey, templatePath)
+	if templateParameters, okType = (*template)[parametersKey].(map[string]interface{}); !okType {
+		return "", errors.Errorf("error asserting data from key \"%s\" in file %q",
+			parametersKey, "azuredeploy.json")
```

Comment:
why not print the path of the file here? (same question for below)

Reply:
we don't have it because we don't rely on file path anymore but on io.Reader for testability.
the error is wrapped from the caller. it's not outputing the whole path though, that can be changed

Reply:
To be clear, is that good enough or do I need to reconstruct the path to the azuredeploy.json?

Reply:
I think it's fine like that, was just curious.

Reply:
added and tested